tests/gnrc_rpl/tests/01-run.py

```diff
@@ -0,0 +1,181 @@
+#!/usr/bin/env python3
+
+# Copyright (C) 2021 Benjamin Valentin
+#
+# This file is subject to the terms and conditions of the GNU Lesser
+# General Public License v2.1. See the file LICENSE in the top level
+# directory for more details.
+
+import subprocess
+import atexit
+import time
+
+from subprocess import Popen
+from riotctrl_shell.gnrc import GNRCICMPv6Echo, GNRCICMPv6EchoParser
+from riotctrl_shell.netif import Ifconfig
+from contextlib import ContextDecorator
+from riotctrl.ctrl import RIOTCtrlBoardFactory
+from riotctrl_ctrl import native
+from riotctrl_shell.netif import IfconfigListParser
+
+PARSERS = {
+    "ping6": GNRCICMPv6EchoParser(),
+    "ifconfig": IfconfigListParser(),
+}
+
+
+class RIOTCtrlAppFactory(RIOTCtrlBoardFactory, ContextDecorator):
+
+    def __init__(self):
+        super().__init__(board_cls={
+            'native': native.NativeRIOTCtrl,
+        })
+        self.ctrl_list = list()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *exc):
+        for ctrl in self.ctrl_list:
+            ctrl.stop_term()
+
+    def get_shell(self, application_directory='.', env={'BOARD': 'native'}):
+        # retrieve a RIOTCtrl Object
+        ctrl = super().get_ctrl(
+            env=env,
+            application_directory=application_directory
+        )
+        # append ctrl to list
+        self.ctrl_list.append(ctrl)
+        # start terminal
+        ctrl.start_term()
+        # return ctrl with started terminal
+        return Shell(ctrl)
+
+    def get_shells(self, num=1):
+        terms = []
+        for i in range(num):
+            terms.append(self.get_shell())
+        return terms
+
+
+class Shell(Ifconfig, GNRCICMPv6Echo):
+    pass
+
+
+def first_netif_and_addr_by_scope(ifconfig_out, scope):
+    netifs = PARSERS["ifconfig"].parse(ifconfig_out)
+    key = next(iter(netifs))
+    netif = netifs[key]
+    return (
+        key,
+        [addr["addr"] for addr in netif["ipv6_addrs"] if addr["scope"] == scope][0],
+    )
+
+
+def global_addr(ifconfig_out):
+    return first_netif_and_addr_by_scope(ifconfig_out, "global")
+
+
+def test_linear_topology(factory):
+
+    # spawn dispatcher
+    zep_dispatch = Popen(['../../dist/tools/zep_dispatch/bin/zep_dispatch',
```

Comment:
The more idiomatic way would be:

```suggestion
    with Popen(['../../dist/tools/zep_dispatch/bin/zep_dispatch',
         '-t', '-', '127.0.0.1', '17754'], stdin = subprocess.PIPE)) as zep_dispatch:
```

which needs the rest of all here indented but would allow you to do away with all the atexit and the manual final termination.

To avoid the indentation, you could take a zep_dispatch as an argument and do nested `with`s in the main part below.

Reply:
Hm with this the test gets stuck on the second iteration - does `zep_dispatch` leaving scope really cause `terminate` being called?

Reply:
I don't quite understand the question.

Anything within a `with` clause gets its exit called when the `with` is left (no matter whether successfully, in an exception or during shutdown).

Whether the main part should be

```python
if __name__ == "__main__":
    with RIOTCtrlAppFactory() as factory:
        with Popen(...) as zep_dispatch:
            test_linear_topology(zep_dispatch, factory)
            test_alternative_route(zep_dispatch, factory)
```

or

```python
if __name__ == "__main__":
    with RIOTCtrlAppFactory() as factory:
        with Popen(...) as zep_dispatch:
            test_linear_topology(zep_dispatch, factory)
        with Popen(...) as zep_dispatch:
            test_alternative_route(zep_dispatch, factory)
```
depends on how zep_dispatch works (but probably the latter, in which case the first process is terminated before the second starts).

Reply:
The problem appears to be that `terminate()` does not get called.

<details><summary>patch</summary>

```patch
diff --git a/tests/gnrc_rpl/tests/01-run.py b/tests/gnrc_rpl/tests/01-run.py
index 77362b998a..7b725f8721 100755
--- a/tests/gnrc_rpl/tests/01-run.py
+++ b/tests/gnrc_rpl/tests/01-run.py
@@ -76,12 +76,7 @@ def global_addr(ifconfig_out):
     return first_netif_and_addr_by_scope(ifconfig_out, "global")


-def test_linear_topology(factory):
-
-    # spawn dispatcher
-    zep_dispatch = Popen(['../../dist/tools/zep_dispatch/bin/zep_dispatch',
-                          '-t', '-', '127.0.0.1', '17754'], stdin=subprocess.PIPE)
-    atexit.register(Popen.terminate, zep_dispatch)
+def test_linear_topology(factory, zep_dispatch):

     # linear topology with 4 nodes
     topology = ("A B\n"
@@ -117,17 +112,8 @@ def test_linear_topology(factory):
     for n in nodes:
         n.stop_term()

-    # terminate dispatcher
-    atexit.unregister(Popen.terminate)
-    zep_dispatch.terminate()
-

-def test_alternative_route(factory):
-
-    # spawn dispatcher
-    zep_dispatch = Popen(['../../dist/tools/zep_dispatch/bin/zep_dispatch',
-                          '-t', '-', '127.0.0.1', '17754'], stdin=subprocess.PIPE)
-    atexit.register(Popen.terminate, zep_dispatch)
+def test_alternative_route(factory, zep_dispatch):

     # topology with 5 nodes with alternative routes
     #   A
@@ -171,12 +157,12 @@ def test_alternative_route(factory):
     for n in nodes:
         n.stop_term()

-    # terminate dispatcher
-    atexit.unregister(Popen.terminate)
-    zep_dispatch.terminate()
-

 if __name__ == "__main__":
     with RIOTCtrlAppFactory() as factory:
-        test_linear_topology(factory)
-        test_alternative_route(factory)
+        with Popen(['../../dist/tools/zep_dispatch/bin/zep_dispatch',
+                    '-t', '-', '127.0.0.1', '17754'], stdin=subprocess.PIPE) as zep_dispatch:
+            test_linear_topology(factory, zep_dispatch)
+        with Popen(['../../dist/tools/zep_dispatch/bin/zep_dispatch',
+                    '-t', '-', '127.0.0.1', '17754'], stdin=subprocess.PIPE) as zep_dispatch:
+            test_alternative_route(factory)
```
</details>

Reply:
My bad -- Popen does implement contexts, but rather than killing the process it awaits it.

A solution can be

```python
if __name__ == "__main__":
    with RIOTCtrlAppFactory() as factory:
        with Popen(...) as zep_dispatch:
            try:
                test_linear_topology(zep_dispatch, factory)
            finally:
                zep_dispatch.terminate()
        with Popen(...) as zep_dispatch:
            try:
                test_alternative_route(zep_dispatch, factory)
            finally:
                zep_dispatch.terminate()
```

but admittedly that's more lines than I'd like it to be. (Ideally Popen would have a flag for that...).

Reply:
thank you, that will work!